### PR TITLE
feat(RateLimits): optionally error on ratelimits

### PR DESCRIPTION
**Pull request by ckohen**
**Please describe the changes this PR makes and why it should be merged:**

Ported from https://github.com/discordjs/discord.js/pull/5659

~~Needs #51~~

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

**Review by vladfrangu — DISMISSED** [review]

**Comment by ckohen** [comment]
Not really sure why CI is failing (cache is set to npm when we're using yarn though, so maybe that's it?)

**Comment by iCrawl** [comment]
That might just be my fault, yes 😆 
Will fix and then request a rebase.

File: packages/rest/src/lib/REST.ts [hunk 1]
```diff
@@ -38,6 +38,14 @@ export interface RESTOptions {
 	 * @default 50
 	 */
 	offset: number;
+	/**
+	 * Determines how how rate limiting and pre-emptive throttling should be handled.
```
Comment (Jiralite) [inline packages/rest/src/lib/REST.ts:42]:
```suggestion
	 * Determines how rate limiting and pre-emptive throttling should be handled.
```

File: packages/rest/src/lib/REST.ts [hunk 2]
```diff
@@ -38,6 +38,14 @@ export interface RESTOptions {
...
+	 * When an array of strings, each element is treated as a prefix for the request route
+	 * (e.g. `/channels/` to match any route starting with `/channels` such as `/channels/:id/messages`)
+	 * for which to throw {@link RateLimitError}s. All other requests routes will be queued normally
```
Comment (Jiralite) [inline packages/rest/src/lib/REST.ts:45]:
```suggestion
	 * for which to throw {@link RateLimitError}s. All other request routes will be queued normally
```

File: packages/rest/src/lib/REST.ts [hunk 3]
```diff
@@ -81,6 +89,10 @@ export interface RateLimitData {
 	 * The bucket hash for this request
 	 */
 	hash: string;
+	/**
+	 * The full url for this request
```
Comment (Jiralite) [inline packages/rest/src/lib/REST.ts:93]:
```suggestion
	 * The full URL for this request
```

File: packages/rest/__tests__/RequestHandler.test.ts [hunk 4]
```diff
@@ -1,8 +1,9 @@
 import nock from 'nock';
-import { DefaultRestOptions, DiscordAPIError, HTTPError, REST, RESTEvents } from '../src';
+import { DefaultRestOptions, DiscordAPIError, HTTPError, RateLimitError, REST, RESTEvents } from '../src';
 
 const api = new REST({ timeout: 2000, offset: 5 }).setToken('A-Very-Fake-Token');
 const invalidAuthApi = new REST({ timeout: 2000 }).setToken('Definitely-Not-A-Fake-Token');
+const rateLimitErrorApi = new REST({ rejectOnRateLimit: ['/channels'] }).setToken('Obviouslly-Not-A-Fake-Token');
```
Comment (kyranet) [inline packages/rest/__tests__/RequestHandler.test.ts:6]:
```suggestion
const rateLimitErrorApi = new REST({ rejectOnRateLimit: ['/channels'] }).setToken('Obviously-Not-A-Fake-Token');
```

**Review by kyranet — DISMISSED** [review]

**Review by kyranet — DISMISSED** [review]

**Review by iCrawl — DISMISSED** [review]

File: packages/rest/src/lib/REST.ts [hunk 5]
```diff
@@ -38,6 +38,14 @@ export interface RESTOptions {
 	 * @default 50
 	 */
 	offset: number;
+	/**
+	 * Determines how rate limiting and pre-emptive throttling should be handled.
+	 * When an array of strings, each element is treated as a prefix for the request route
+	 * (e.g. `/channels/` to match any route starting with `/channels` such as `/channels/:id/messages`)
```
Comment (vladfrangu) [inline packages/rest/src/lib/REST.ts:44]:
```suggestion
	 * (e.g. `/channels` to match any route starting with `/channels` such as `/channels/:id/messages`)
```

File: packages/rest/src/lib/handlers/SequentialHandler.ts [hunk 6]
```diff
@@ -356,6 +379,29 @@ export class SequentialHandler {
 					`  Retry After    : ${retryAfter}ms`,
 				].join('\n'),
 			);
+			const isGlobal = this.globalLimited;
+			let limit: number;
+			let timeout: number;
+
+			if (isGlobal) {
+				// Set RateLimitData based on the globl limit
```
Comment (vladfrangu) [inline packages/rest/src/lib/handlers/SequentialHandler.ts:387]:
```suggestion
				// Set RateLimitData based on the global limit
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Review by iCrawl — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**